### ISSUE TITLE
Ability to apply dynamic QueryFilter again after changing property

### ISSUE THREAD

**Issue by Saibamen** [dotnet/efcore#23718]
## Ask a question

Hello.

My team are re-creating backend solution with EF Core 5 based on EF6 code.
This is multi-tenant application.

We have filters that are applied in `OnModelCreating()`.

Yesterday I noticed that some of our controllers creating DBContext in constructors and setting `CurrentTenantId` in controller actions, which will never works, because our filters are applied only once (on model creating):

```C#
private readonly MyContext _dbTenant;

public MyController(IMyContextFactory contextFactory) : base(contextFactory)
{
    _dbTenant = ContextFactory.GetMyContext();
}

public IActionResult PostSomething(Guid guid)
{
    _dbTenant.CurrentTenantId = UserToken.TenantId;

    var repo = new MyRepository(_dbTenant)
}
```

Switching `ModelBuilder` into `IModel` in my QueryFilter will not work here, because at the end I could not do `SetQueryFilter()`.

I will set setter for `CurrentTenantId` in `MyContext` as private and create context in controllers actions `ContextFactory.GetMyContext(UserToken.TenantId)`.
I have only 25 places, where we are changing `CurrentTenantId` in already created Context, so this is not a big deal to change this, but I am wondering if there is some better way to fix this problem?
Calling somehow `ApplyFilters()` again if `CurrentTenantId` will change? I remember that someone from EF Team (@ajcvickers maybe) told me we cannot use `ModelBuilder` again after model has been created.

### Include your code

`TenantQueryFilter`:

```C#
private readonly long _tenantId;

public TenantQueryFilter(long tenantId)
{
    _tenantId = tenantId;
}

public void Apply(ModelBuilder modelBuilder)
{
    if (_tenantId <= 0)
    {
        return;
    }

    // generic way to achieve this code:
    // modelBuilder.Entity<MyEntity>().HasQueryFilter(e => EF.Property<long>(e, "TenantId") == TenantId);

    // get EF entities with TenantAttribute
    var entitiesWithTenantAttribute = GetEntitiesWithAttribute(modelBuilder);

    foreach (var entityType in entitiesWithTenantAttribute)
    {
        var colName = GetColumnName(entityType);
        if (colName == null)
        {
            continue;
        }

        // build filter: (p) => p.{colName} == {CurrentTenantId}
        // i.e. (p) => p.TenantId == 123)
        var parameter = Expression.Parameter(entityType.ClrType, "p");
        var filter = Expression.Lambda(
            Expression.Equal(
                Expression.Property(parameter, colName),
                Expression.Constant(_tenantId)
            ),
            parameter);

        entityType.SetQueryFilter(filter);
    }
}

public IEnumerable<IMutableEntityType> GetEntitiesWithAttribute(ModelBuilder modelBuilder)
{
    var entitiesWithAttribute = modelBuilder.Model
        .GetEntityTypes()
        .Where(e => e.ClrType.GetCustomAttribute(typeof(T), false) != null);
    return entitiesWithAttribute;
}
```

Applying filters in `OnModelCreating`:
```C#
protected override void OnModelCreating(ModelBuilder modelBuilder)
{
    new MyDbModelBuilder(modelBuilder).ApplyConfigurations();

    // NOTE: thanks to our custom DynamicModelCacheKeyFactory we have OnModelCreating() not cached
    // but called also for new instances (i.e. with different CurrentTenantId)
    // which is important to correctly set filters (in this case: TenantQueryFilter)
    ApplyFilters(modelBuilder);
}
```

```C#
private void ApplyFilters(ModelBuilder modelBuilder)
{
    var filters = new List<IQueryFilter>()
    {
        new TenantQueryFilter(CurrentTenantId),
        new SoftDeleteQueryFilter(),
    };

    foreach (var filter in filters)
    {
        filter.Apply(modelBuilder);
    }
}
```

### Include provider and version information

EF Core version: 5.0.1
Database provider: Microsoft.EntityFrameworkCore.SqlServer
Target framework: .NET Core 3.1
Operating system:
IDE: Visual Studio 2019 16.8.3

**Comment by smitpatel** [comment]
> Expression.Constant(_tenantId)

Using constant will capture the value of _tenantId at the time OnModelCreating is run. What you need is to use is `context.CurrentTenantId` in the query filter definition (where context is your derived DbContext). If you configure it that way, EF Core will use the value of CurrentTenantId from the db context which is executing the query automatically.

**Comment by Saibamen** [comment]
Good point @smitpatel, but even if I use `Expression.Constant(myContext.CurrentTenantId)` instead of `Expression.Constant(_tenantId)` and call `filter.Apply(modelBuilder, this);` inside `ApplyFilters(ModelBuilder modelBuilder)`, this `ApplyFilters()` will be called only once (in `OnModelCreating()`).

**Comment by smitpatel** [comment]
The key is you don't use Expression.Constant.
Since you are creating your expression tree yourself a straightforward way would be to use `Expression.MakeMemberAccess(Expression.Constant(new MyContext()), "CurrentTenantId")`

**Comment by Saibamen** [comment]
Thanks Smit!

Final solution:
```C#
var currentTenantIdMemberInfo = myContext.GetType().GetMember(nameof(myContext.CurrentTenantId)).FirstOrDefault();

if (currentTenantIdMemberInfo == null)
{
    throw new NullReferenceException("Cannot get MemberInfo for CurrentTenantId from Context");
}
```
```C#
Expression.MakeMemberAccess(Expression.Constant(myContext), currentTenantIdMemberInfo)
```

And of course I needed to get rid of `if (_tenantId <= 0) return`

I have another issue, because I want to not apply filter (just create lambda `(p => p)`) if `CurrentTenantId` <= 0, but I already asked this question in SO: https://stackoverflow.com/questions/65366388/lambda-ifthenelse-expression-with-another-lambda-inside-for-setqueryfilter

**Comment by smitpatel** [comment]
Easy fix
So your current filter is in this form right now
```C#
() => IIF((value(My.Project.MyContext).CurrentTenantId > 0), () => (p.TenantId == value(My.Project.MyContext).CurrentTenantId), () => p)
```
Which is basically translate to following code block in C#
```C#
If (context.CurrentTenantId > 0)
{
    return () => p.TenantId == context.CurrentTenantId;
}
else
{
    return () => p
}
```
Which is invalid since the definition of filter is supposed to be `Func<TEntity, bool>` Above has missing `p` the entity parameter and it also returns a func. It worked since the method signature requires a lambda expression only.

Now what you really want is this
```C#
p => context.CurrentTenantId > 0 ? p.TenantId == context.CurrentTenantId : true
```
A better way to write using short-circuiting of logical operation is
```C#
p => context.CurrentTenantId <= 0 || p.TenantId == context.CurrentTenantId
```
The expression tree for it would be (on-the-fly, forgive minor compilation errors if any)
```C#
Expression.Lambda(
    Expression.OrElse(
        Expression.LessThanOrEqual(
            contextMemberAccess, 
            Expression.Constant((long)0)),
        Expression.Equal(
            Expression.Property(parameter, colName),
            contextMemberAccess)),
    parameter);
```

P.S. - Expression has 2 constructs for condition scenarios. Expression.Condition method creates a ternary expression which returns a value. Expression.IfThenElse method creates a if/else block which may not always have to return. While both are somewhat interchangeable through block expression and return labels, IfThenElse is generally used when there is nothing to return (like a function returning void).

**Comment by Saibamen** [comment]
Awesome Smit!

Thank you very much with this lambda. Now everything works as it was supposed to.

Merry Christmas and happy New Year.

**Comment by mtapis** [comment]
I have a similar problem about this situation , my project does work for different tenants like it should but i can not add migration because i set properties in TenantDbContext constructor
```c#
    public static class ModelBuilderExtensions
    {
        public static void AddGlobalTenantIdFilter(this ModelBuilder modelBuilder, TenantDbContext dbContext)
        {
            //Global filter TenantId alanı null olmayan ve basetype alanı null olan entityler için uygulanır.
            var entitiesForFilter = modelBuilder.Model.GetEntityTypes()
                .Where(x => x.BaseType == null && x.FindProperty(CustomClaimTypes.TenantId) != null).ToList();

            foreach (var entityType in entitiesForFilter)
            {
                var isTenantIdProperty = entityType.FindProperty(CustomClaimTypes.TenantId);

                //Seedlenen kayıtlara erişilebilmesi için tenantId == context.currenttTenantId || tenantId == null sorgusu eklenecek
                if (isTenantIdProperty != null && isTenantIdProperty.ClrType == typeof(Guid?))
                {
                    var parameter = Expression.Parameter(entityType.ClrType, "x");
                    var tenantIdProperty = Expression.Property(parameter, CustomClaimTypes.TenantId);
                    //dbcontext.CurrentTenantId değerini sorguya ekliyoruz TenantId değeri dinamik olarak setlenmeli.
                    var currentTenantIdMemberInfo = dbContext.GetType().GetMember(nameof(dbContext.CurrentTenantId)).FirstOrDefault();
                   
                    if (currentTenantIdMemberInfo != null)
                    {
                        //dbcontext.CurrentTenantId değerine erişim sağlayan bir expression oluşturulur.
                        var tenantIdValueExpression = Expression.MakeMemberAccess(Expression.Constant(dbContext), currentTenantIdMemberInfo);
                        var tenantIdFilter = Expression.Equal(tenantIdProperty, tenantIdValueExpression);

                        //tenantId == null expressionu oluşturulur
                        var tenantIdNullFilter = Expression.Equal(tenantIdProperty, Expression.Constant(null, typeof(Guid?)));
                        var combinedFilter = Expression.OrElse(tenantIdFilter, tenantIdNullFilter);

                        var lambda = Expression.Lambda(combinedFilter, parameter);
                        modelBuilder.Entity(entityType.ClrType).HasQueryFilter(lambda);
                    }
                    
                }

               
            }
        }
    }
``` 
also i add query filters base on the tenant information that i get from HttpContext 
```c#
    protected override void OnModelCreating(ModelBuilder modelBuilder)
    {
        modelBuilder.UseCollation(_dbUtilities.GetDBCollationDefinition());
        modelBuilder.ApplyConfigurationsFromAssembly(Assembly.GetExecutingAssembly());
        SeedDatabase(modelBuilder);
        //Elle migration yapıldığında contextData içi boş gelecek. Bu nedenle _contextData?.UserId kontrolü eklendi.
        //DBMigrator tarafından çağrıldığında contextData null geliyor.
        if (_contextData == null || string.IsNullOrEmpty(_contextData?.UserId))
        {
            base.OnModelCreating(modelBuilder);
            return;
        }
        if (CurrentSingleDbMode)
        {
            modelBuilder.AddGlobalTenantIdFilter(this);
        }
        base.OnModelCreating(modelBuilder);

    }

``` 
```c#
public class TenantDbContext : IdentityDbContext<AspNetUser, AspNetRole, Guid,
    AspNetUserClaim, AspNetUserRole, AspNetUserLogin, AspNetRoleClaim, AspNetUserToken>
{
    private readonly IConfiguration _configuration;
    private readonly IHttpContextData _contextData;
    private readonly IDbUtilities _dbUtilities;
    private readonly IMemoryCache _memoryCache;
    private readonly ICryptography _cryptography;
    private readonly ISharedLocalizer _sharedLocalizer;
    private readonly SeedStatikReferanslar _seedStatikReferanslar;
    private readonly SeedRefTip _seedSystemReferanslar;
    private readonly SeedReferanslar _seedReferanslar;

    public TenantDbContext(DbContextOptions<TenantDbContext> options, IMemoryCache memoryCache,
        IHttpContextData contextData, ICryptography cryptography,IConfiguration configuration,
        IDbUtilities dbUtilities, ISharedLocalizer sharedLocalizer,
        SeedStatikReferanslar seedStatikReferanslar, SeedRefTip seedSystemReferanslar, SeedReferanslar seedReferanslar,
        bool DoNotSeedDB = false) : base(options)
    {
        _memoryCache = memoryCache;
        _contextData = contextData;
        _seedStatikReferanslar = seedStatikReferanslar;
        _configuration = configuration;
        _dbUtilities = dbUtilities;
        _seedSystemReferanslar = seedSystemReferanslar;
        _seedReferanslar = seedReferanslar;
        DoNotSeed = DoNotSeedDB;
        _cryptography = cryptography;
        _sharedLocalizer = sharedLocalizer;
        CurrentTenantId = contextData.TenantId;
        CurrentSingleDbMode = contextData.SingleDbMode;

    }
    public Guid? CurrentTenantId { get; set; }
    public bool CurrentSingleDbMode { get; set; }
    internal string ConnectionString { get; set; }
    public bool DoNotSeed { get; set; } = false;
}
```
OnModelCreating code block runs for once the problem is queryfilters must work base on tenant so if the tenant has singleDBMode then i  should apply filters

**Comment by roji** [comment]
@mtapis it's very much by design that OnModelCreating runs only once - anything else would be extremely inefficient.

Query filters do support varying a parameter, such as a tenant ID - [see the explanation and code sample in our docs](https://learn.microsoft.com/en-us/ef/core/querying/filters). In problem in your above code is likely that you're trying to reference the DbContext instance as a constant (`Expression.Constant(dbContext)`); it's up to you to generate the correct tree structure when manually building up expression trees.

In a cursory glance, the above doesn't look like it requires manual building up of expression trees at all (with Expression) - you may be able to just write it using regular C# code and have the compiler generate the tree for you, as usual. Otherwise I advise looking at the exact tree generated by the compiler and creating such a tree yourself.

**Comment by mtapis** [comment]
The problem is  when i pass the contextdata.TenantId value inside the query filters , it detects the first request tenant information and sets the queryfilters with that tenantId , but when I made another request with different tenant it still shows me the first tenants data
Code Below for that 
```c#
    protected override void OnModelCreating(ModelBuilder modelBuilder)
    {
        modelBuilder.UseCollation(_dbUtilities.GetDBCollationDefinition());
        modelBuilder.ApplyConfigurationsFromAssembly(Assembly.GetExecutingAssembly());
        SeedDatabase(modelBuilder);
        //Elle migration yapıldığında contextData içi boş gelecek. Bu nedenle _contextData?.UserId kontrolü eklendi.
        //DBMigrator tarafından çağrıldığında contextData null geliyor.
        if (_contextData == null || string.IsNullOrEmpty(_contextData?.UserId))
        {
            base.OnModelCreating(modelBuilder);
            return;
        }
        if (_contextData.SingleDbMode)
        {
            modelBuilder.AddGlobalTenantIdFilter(_contextData.TenantId);
        }
        base.OnModelCreating(modelBuilder);

    }
``` 
_contextData is my Service that uses IHttpContextAccessor and set tenant information from requests , I checked the MultiTenancy documentation , there is _tenant defined in DbContext class and its been set inside constructor using HttpContext, the problem is if developer tries to add migration from powershell HttpContext will be null so _tenant=context.Tenant should not work; it does not work in my case , Thank you for your help Roji !

**Comment by roji** [comment]
I'd advise trying to follow the guidance in the docs on this: inject the tenant ID into your DbContext directly (as a private field), and then reference that field from the query filter - that should work. Also, avoid manually constructing expression trees via Expression - that's quite advanced and error-prone, and I'm not sure I see a need for that for your scenario.
